text/0097-log-data-model.md

```diff
@@ -0,0 +1,972 @@
+by a logging system.
+
+## Design Notes
+
+### Requirements
+
+The Data Model was designed to satisfy the following requirements:
+
+- It should be possible to unambiguously map existing log formats to this Data
+  Model. Translating log data from an arbitrary log format to this Data Model
+  and back should ideally result in identical data.
+
+- Mappings of other log formats to this Data Model should be semantically
+  meaningful. The Data Model must preserve the semantics of particular elements
+  of existing log formats.
+
+- Translating log data from an arbitrary log format A to this Data Model and
+  then translating from the Data Model to another log format B ideally must
+  result in a meaningful translation of log data that is no worse than a
+  reasonable direct translation from log format A to log format B.
+
+- It should be possible to efficiently represent the Data Model in concrete
+  implementations that require the data to be stored or transmitted. We
+  primarily care about 2 aspects of efficiency: CPU usage for
+  serialization/deserialization and space requirements in serialized form. This
+  is an indirect requirement that is affected by the specific representation of
+  the Data Model rather than the Data Model itself, but is still useful to keep
+  in mind.
+
+The Data Model aims to successfully represent 3 sorts of logs and events:
+
+- System Formats. These are logs and events generated by the operating system
+  and over which we have no control - we cannot change the format or affect what
+  information is included (unless the data is generated by an application which
+  we can modify). An example of system format is Syslog.
+
+- Third-party Applications. These are generated by third-party applications. We
+  may have certain control over what information is included, e.g. customize the
+  format. An example is Apache log file.
+
+- First-party Applications. These are applications that we develop and we have
+  some control over how the logs and events are generated and what information
+  we include in the logs. We can likely modify the source code of the
+  application if needed.
+
+### Field Kinds
+
+This Data Model defines a logical model for a log record (irrespective of the
+physical format and encoding of the record). Each record contains 2 kinds of
+fields:
+
+- Named top-level fields of specific type and meaning.
+
+- Fields stored in the key/value pair lists, which can contain arbitrary values
+  of different types. The keys and values for well-known fields follow semantic
+  conventions for key names and possible values that allow all parties that work
+  with the field to have the same interpretation of the data. See references to
+  semantic conventions for `Resource` and `Attributes` fields and examples in
+  [Appendix A](#appendix-a-example-mappings).
+
+The reasons for having these 2 kinds of fields are:
+
+- Ability to efficiently represent named top-level fields, which are almost
```

Comment:
will we enforce the "type" or "name" of the record to know what set of named fields to expect?

Reply:
It depends on the concrete representation of this data model. For example when represented in ProtoBuf the top-level fields will be likely represented as ProtoBuf message fields. This means field type will be enforced. The names will not be enforced on the wire (since ProtoBuf does not transmit field names on the wire), but they will be enforced by compiler (in compiled languages) and at runtime when encoding/decoding (for dynamic languages).

For other formats the answer may be different. For example we may have a JSON representation where neither type nor name of the fields are enforced on the wire.

Reply:
Closing, assuming answered. Feel free to reopen if needed.